### PR TITLE
Fix a selection of `const` related issues with `ranget`.

### PR DESCRIPTION
**Pull request by thomasspriggs**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

Attempting to use `ranget` on const input collections or attempting to use a `const ranget` would previously cause compile errors. This PR fixes these issues.

### CODE REVIEW

**Comment by kroening** [comment]
I would squash this slightly more, e.g., put the test into the commit of the change.

**Comment by kroening** [comment]
Also, I'd change the return types of
```
const value_type &operator*() const
const value_type *operator->() const
```
for consistency.

**Review by kroening — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7a5ff904147f020a431e67193346e4108caaca3c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97863971

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 20cc2dd423d0a5b5b8281cb663524c9664bbdfea).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97872239

File: unit/util/range.cpp [hunk 1]
```diff
@@ -42,16 +50,25 @@ SCENARIO("range tests", "[core][util][range]")
...
     {
-      auto filtered_range = make_range(list).filter(
+      const auto filtered_range = make_range(list).filter(
         [&](const std::string &s) { return s.length() == 4; });
       auto it = filtered_range.begin();
       REQUIRE(*it == "cdef");
       ++it;
       REQUIRE(it == filtered_range.end());
     }
+    THEN("A const filter iterator can mutate the input collection.")
```
Comment (smowton) [inline unit/util/range.cpp]: Suggest rephrasing, since "const filter iterator" sounds an awful lot like a `const_iterator`, which this is not

**Review by smowton — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a89ca18e34f3098ceebb2756122454994306564b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/97929087

**Pull request merged**